File: backend/models/Order.js
const mongoose = require('mongoose');

const orderItemSchema = new mongoose.Schema({
  medicine: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Medicine',
    required: true
  },
  quantity: {
    type: Number,
    required: true,
    min: [1, 'Quantity must be at least 1']
  },
  price: {
    type: Number,
    required: true,
    min: [0, 'Price cannot be negative']
  },
  pharmacy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Pharmacy',
    required: true
  }
});

const orderSchema = new mongoose.Schema({
  orderId: {
    type: String,
    unique: true
  },
  customer: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  // Customer information for orders without authentication
  customerName: {
    type: String,
    trim: true
  },
  customerEmail: {
    type: String,
    trim: true,
    lowercase: true
  },
  customerPhone: {
    type: String,
    trim: true
  },
  pharmacy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Pharmacy',
    required: true
  },
  items: [orderItemSchema],
  status: {
    type: String,
    enum: ['pending', 'confirmed', 'processing', 'ready', 'delivering', 'delivered', 'cancelled'],
    default: 'pending'
  },
  totalAmount: {
    type: Number,
    required: true,
    min: [0, 'Total amount cannot be negative']
  },
  deliveryAddress: {
    type: String,
    required: true,
    trim: true
  },
  deliveryLatitude: {
    type: Number
  },
  deliveryLongitude: {
    type: Number
  },
  prescriptionUrl: {
    type: String,
    trim: true
  },
  specialInstructions: {
    type: String,
    trim: true,
    maxlength: [500, 'Instructions cannot be more than 500 characters']
  },
  paymentMethod: {
    type: String,
    enum: ['cash', 'card', 'online'],
    default: 'cash'
  },
  paymentStatus: {
    type: String,
    enum: ['pending', 'paid', 'failed', 'refunded'],
    default: 'pending'
  },
  driver: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  estimatedDeliveryTime: {
    type: Date
  },
  actualDeliveryTime: {
    type: Date
  },
  trackingHistory: [{
    status: {
      type: String,
      enum: ['pending', 'confirmed', 'processing', 'ready', 'delivering', 'delivered', 'cancelled']
    },
    timestamp: {
      type: Date,
      default: Date.now
    },
    note: {
      type: String,
      trim: true
    }
  }],
  rating: {
    type: Number,
    min: 1,
    max: 5
  },
  review: {
    type: String,
    trim: true,
    maxlength: [500, 'Review cannot be more than 500 characters']
  }
}, {
  timestamps: true
});

// Indexes for better query performance
orderSchema.index({ customer: 1, createdAt: -1 });
orderSchema.index({ pharmacy: 1, createdAt: -1 });
orderSchema.index({ driver: 1, status: 1 });
orderSchema.index({ orderId: 1 });
orderSchema.index({ status: 1 });

// Pre-save middleware to generate orderId
orderSchema.pre('save', function(next) {
  if (this.isNew && !this.orderId) {
    const timestamp = Date.now().toString().slice(-6);
    const random = Math.floor(Math.random() * 1000).toString().padStart(3, '0');
    this.orderId = `ORD-${timestamp}-${random}`;
  }
  next();
});

// Virtual for formatted total
orderSchema.virtual('formattedTotal').get(function() {
  return `TND ${this.totalAmount.toFixed(2)}`;
});

// Method to update status with tracking history
orderSchema.methods.updateStatus = function(newStatus, note = '') {
  this.status = newStatus;
  this.trackingHistory.push({
    status: newStatus,
    note: note
  });

  // Set delivery time when delivered
  if (newStatus === 'delivered') {
    this.actualDeliveryTime = new Date();
  }
};

// Method to assign driver
orderSchema.methods.assignDriver = function(driverId) {
  this.driver = driverId;
  this.status = 'confirmed';
  this.trackingHistory.push({
    status: 'confirmed',
    note: 'Driver assigned'
  });
};

// Static method to get orders by status
orderSchema.statics.getOrdersByStatus = function(status) {
  return this.find({ status }).populate('customer pharmacy driver');
};

// Ensure virtual fields are serialized
orderSchema.set('toJSON', { virtuals: true });

module.exports = mongoose.model('Order', orderSchema);

File: app/src/main/java/tn/rnu/isetr/miniprojet/data/Order.kt
package tn.rnu.isetr.miniprojet.data

import tn.rnu.isetr.miniprojet.data.Pagination

data class OrderRequest(
    val pharmacy: String, // pharmacy ID
    val items: List<OrderItem>,
    val deliveryAddress: String,
    val deliveryLatitude: Double? = null,
    val deliveryLongitude: Double? = null,
    val specialInstructions: String? = null,
    val prescriptionUrl: String? = null
)

data class OrderItem(
    val medicine: String, // medicine ID
    val quantity: Int
)

data class OrderResponse(
    val success: Boolean,
    val message: String,
    val data: Order? = null,
    val errors: List<String>? = null
)

data class Order(
    val _id: String,
    val customer: String?, // Customer ID or name depending on population
    val pharmacy: Pharmacy,
    val items: List<OrderItemDetail>,
    val totalAmount: Double,
    val status: String,
    val deliveryAddress: String,
    val deliveryLatitude: Double? = null,
    val deliveryLongitude: Double? = null,
    val specialInstructions: String? = null,
    val prescriptionUrl: String? = null,
    val createdAt: String,
    val updatedAt: String
)

data class OrderItemDetail(
    val medicine: Medicine,
    val quantity: Int,
    val price: Double,
    val pharmacy: String
)

data class OrderListResponse(
    val success: Boolean,
    val data: List<Order>,
    val pagination: Pagination? = null,
    val message: String? = null
)

data class UploadResponse(
    val success: Boolean,
    val message: String,
    val data: UploadData
)

data class UploadData(
    val url: String,
    val filename: String
)
